File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/GitLogNumstatRawParserStrategy.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogLineCollector;
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy;

import java.time.LocalDateTime;
import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collector;
import java.util.stream.Stream;

import static de.maibornwolff.codecharta.importer.scmlogparser.parser.git.AuthorParser.AUTHOR_ROW_INDICATOR;
import static de.maibornwolff.codecharta.importer.scmlogparser.parser.git.CommitDateParser.DATE_ROW_INDICATOR;

public class GitLogNumstatRawParserStrategy implements LogParserStrategy {

    public static final String CORRESPONDING_LOG_CREATION_CMD = "git log --numstat --raw --topo-order";
    public static final java.util.function.BiFunction<Modification, Modification, Modification> MODIFICATION_MODIFICATION_MODIFICATION_BI_FUNCTION = (a, b) -> a;
    private static final Predicate<String> GIT_COMMIT_SEPARATOR_TEST = logLine -> logLine.startsWith("commit");
    private static final String FILE_LINE_SPLITTER = "\\s+";

    private static final BinaryOperator<Map<String, Modification>> throwUnsupportedOperationException =
            (a, b) -> {
                throw new UnsupportedOperationException("parallel collection lines not supported");
            };
    private static final Function<Map<String, Modification>, ArrayList<Modification>> mapValuesToList = map -> new ArrayList<>(map.values());

    private static boolean isFileLine(String commitLine) {
        return GitLogRawParserStrategy.isFileLine(commitLine) || GitLogNumstatParserStrategy.isFileLine(commitLine);
    }

    static Modification parseModification(String fileLine) {
        if (fileLine.startsWith(":")) {
            return GitLogRawParserStrategy.parseModification(fileLine);
        }

        return GitLogNumstatParserStrategy.parseModification(fileLine);
    }

    private static Modification mergeModifications(Modification... a) {
        String filename = a[0].getFilename();
        int additions = Arrays.stream(a).mapToInt(Modification::getAdditions).sum();
        int deletions = Arrays.stream(a).mapToInt(Modification::getDeletions).sum();
        Modification.Type type = Arrays.stream(a)
                .map(Modification::getType)
                .filter(t -> t != Modification.Type.UNKNOWN)
                .findFirst()
                .orElse(Modification.Type.UNKNOWN);

        if (type == Modification.Type.RENAME) {
            String oldFilename = Arrays.stream(a)
                    .map(Modification::getOldFilename)
                    .filter(s -> !s.isEmpty())
                    .findFirst()
                    .orElse("");
            return new Modification(filename, oldFilename, additions, deletions, type);
        }

        return new Modification(filename, additions, deletions, type);
    }

    @Override
    public List<String> listSupportedMetrics() {
        return Arrays.asList(
                "code_churn",
                "loc",
                "number_of_authors",
                "number_of_commits",
                "weeks_with_commits"
        );
    }

    @Override
    public Collector<String, ?, Stream<List<String>>> createLogLineCollector() {
        return LogLineCollector.create(GIT_COMMIT_SEPARATOR_TEST);
    }

    @Override
    public Optional<String> parseAuthor(List<String> commitLines) {
        return commitLines.stream()
                .filter(commitLine -> commitLine.startsWith(AUTHOR_ROW_INDICATOR))
                .map(AuthorParser::parseAuthor)
                .findFirst();

    }

    @Override
    public List<Modification> parseModifications(List<String> commitLines) {
        BiConsumer<Map<String, Modification>, Modification> accumulator =
                (map, mod) -> map.merge(mod.getFilename(), mod, GitLogNumstatRawParserStrategy::mergeModifications);


        return commitLines.stream()
                .filter(GitLogNumstatRawParserStrategy::isFileLine)
                .map(GitLogNumstatRawParserStrategy::parseModification)
                .collect(Collector.of(HashMap::new, accumulator, throwUnsupportedOperationException, mapValuesToList));
    }

    @Override
    public Optional<LocalDateTime> parseDate(List<String> commitLines) {
        return commitLines.stream()
                .filter(commitLine -> commitLine.startsWith(DATE_ROW_INDICATOR))
                .map(CommitDateParser::parseCommitDate)
                .findFirst();
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/git/GitLogNumstatRawParserStrategy.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser.git

import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogLineCollector
import de.maibornwolff.codecharta.importer.scmlogparser.parser.LogParserStrategy
import de.maibornwolff.codecharta.importer.scmlogparser.parser.git.AuthorParser.AUTHOR_ROW_INDICATOR
import de.maibornwolff.codecharta.importer.scmlogparser.parser.git.CommitDateParser.DATE_ROW_INDICATOR
import java.time.OffsetDateTime
import java.util.function.Predicate
import java.util.stream.Collector
import java.util.stream.Stream
import kotlin.streams.toList

class GitLogNumstatRawParserStrategy : LogParserStrategy {
    override fun creationCommand(): String {
        return "git log --numstat --raw --topo-order"
    }

    override fun createLogLineCollector(): Collector<String, *, Stream<List<String>>> {
        return LogLineCollector.create(GIT_COMMIT_SEPARATOR_TEST)
    }

    override fun parseAuthor(commitLines: List<String>): String {
        return commitLines
            .parallelStream()
            .filter { commitLine -> commitLine.startsWith(AUTHOR_ROW_INDICATOR) }
            .map { AuthorParser.parseAuthor(it) }
            .toList()
            .first()
    }

    override fun parseModifications(commitLines: List<String>): List<Modification> {
        return commitLines
            .filter { isFileLine(it) }
            .map { parseModification(it) }
            .groupingBy { it.filename }
            .aggregate { _, aggregatedModification: Modification?, currentModification, _ ->
                when (aggregatedModification) {
                    null -> mergeModifications(currentModification)
                    else -> mergeModifications(aggregatedModification, currentModification)
                }
            }
            .values
            .filterNotNull()
            .toList()
    }

    override fun parseDate(commitLines: List<String>): OffsetDateTime {
        return commitLines
            .filter { commitLine -> commitLine.startsWith(DATE_ROW_INDICATOR) }
            .map { CommitDateParser.parseCommitDate(it) }
            .first()
    }

    companion object {
        private val GIT_COMMIT_SEPARATOR_TEST = Predicate<String> { logLine -> logLine.startsWith("commit") }
        private fun isFileLine(commitLine: String): Boolean {
            return GitLogRawParserStrategy.isFileLine(commitLine) || GitLogNumstatParserStrategy.isFileLine(commitLine) }

        internal fun parseModification(fileLine: String): Modification {
            return if (fileLine.startsWith(":")) {
                GitLogRawParserStrategy.parseModification(fileLine)
            } else GitLogNumstatParserStrategy.parseModification(fileLine)
        }

        private fun mergeModifications(vararg a: Modification): Modification {
            val filename = a[0].filename
            val additions = a.map { it.additions }.sum()
            val deletions = a.map { it.deletions }.sum()
            val type =
                a.map { it.type }.firstOrNull { t -> t != Modification.Type.UNKNOWN } ?: Modification.Type.UNKNOWN

            if (type == Modification.Type.RENAME) {
                val oldFilename = a.map { it.oldFilename }.firstOrNull { s -> s.isNotEmpty() } ?: ""
                return Modification(filename, oldFilename, additions, deletions, type)
            }

            return Modification(filename, additions, deletions, type)
        }
    }
}
